sender_test.go

```diff
@@ -2,32 +2,769 @@ package amqp
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+
+	sendInitialFlowFrame(t, netConn, 0, 100)
+	time.Sleep(100 * time.Millisecond)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	msg := NewMessage([]byte("test"))
+	msg.DeliveryTag = make([]byte, 33)
+	require.Error(t, snd.Send(ctx, msg))
+	cancel()
+
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, client.Close())
+}
+
+func TestSenderSendMultiTransfer(t *testing.T) {
+	var deliveryID uint32
+	transferCount := 0
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.EncodeFrame(mocks.FrameAMQP, 0, &frames.PerformOpen{
+				ChannelMax:   65535,
+				ContainerID:  "container",
+				IdleTimeout:  time.Minute,
+				MaxFrameSize: 128, // really small max frame size
+			})
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			return mocks.SenderAttach(0, tt.Name, 0, encoding.ModeUnsettled)
+		case *frames.PerformTransfer:
+			if tt.DeliveryID != nil {
+				// deliveryID is only sent on the first transfer frame for multi-frame transfers
+				if transferCount != 0 {
+					return nil, fmt.Errorf("unexpected DeliveryID for frame number %d", transferCount)
+				}
+				deliveryID = *tt.DeliveryID
+			}
+			if tt.MessageFormat != nil && transferCount != 0 {
+				// MessageFormat is only sent on the first transfer frame for multi-frame transfers
+				return nil, fmt.Errorf("unexpected MessageFormat for frame number %d", transferCount)
+			} else if tt.MessageFormat == nil && transferCount == 0 {
+				return nil, errors.New("unexpected nil MessageFormat")
+			}
+			if tt.More {
+				transferCount++
+				return nil, nil
+			}
+			return mocks.PerformDisposition(encoding.RoleReceiver, 0, deliveryID, &encoding.StateAccepted{})
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+
+	sendInitialFlowFrame(t, netConn, 0, 100)
+	time.Sleep(100 * time.Millisecond)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100000*time.Millisecond)
+	payload := make([]byte, 512)
```

Comment:
Just to make the link, it might be nice to use a const for the max frame size of this test, and do the math inline here so we can see how they relate.